x-pack/lib/filters/geoip/database_manager.rb

```diff
@@ -30,134 +31,193 @@ module LogStash module Filters module Geoip class DatabaseManager
+                                               Concurrent::Array.new,
+                                               city_database_path,
+                                               cc_city_database_path),
+                "#{ASN}" => DatabaseState.new(@metadata.is_eula(ASN),
+                                              Concurrent::Array.new,
+                                              asn_database_path,
+                                              cc_asn_database_path) }
+
+    @download_manager = DownloadManager.new(@metadata)
+
+    @trigger_download = Concurrent::AtomicBoolean.new(false)
   end
 
+  protected
+  # notice plugins to update database path to the new download
+  # update timestamp when download is valid or there is no update
+  # do daily check and clean up
   def execute_download_job
     begin
-      has_update, new_database_path = @download_manager.fetch_database
-      @database_path = new_database_path if has_update
-      @metadata.save_timestamp(@database_path)
-      has_update
+      updated_db = @download_manager.fetch_database
+      updated_db.each do |database_type, valid_download, dirname, new_database_path|
+        if valid_download
+          @metadata.save_metadata(database_type, dirname, true)
+          @states[database_type].is_eula = true
+          @states[database_type].database_path = new_database_path
+          @states[database_type].plugins.dup.each { |plugin| plugin.update_database(new_database_path) if plugin }
+        end
+      end
+
+      updated_type = updated_db.map { |database_type, valid_download, new_database_path| database_type }
+      (DB_TYPES - updated_type).each { |unchange_type| @metadata.update_timestamp(unchange_type) }
     rescue => e
       logger.error(e.message, :cause => e.cause, :backtrace => e.backtrace)
+    ensure
       check_age
-      false
+      clean_up_database
+    end
+  end
+
+  # call expiry action if database is expired and EULA
+  def check_age(database_types = DB_TYPES)
+    database_types.map do |database_type|
+      days_without_update = (::Date.today - ::Time.at(@metadata.updated_at(database_type)).to_date).to_i
+
+      case
+      when days_without_update >= 30
+        if @states[database_type].is_eula && @states[database_type].plugins.size > 0
+          logger.error("The MaxMind database hasn't been updated from last 30 days. Logstash is unable to get newer version from internet. "\
+            "According to EULA, GeoIP plugin needs to stop using MaxMind database in order to be compliant. "\
+            "Please check the network settings and allow Logstash accesses the internet to download the latest database, "\
+            "or switch to offline mode (:database => PATH_TO_YOUR_DATABASE) to use a self-managed database "\
+            "which you can download from https://dev.maxmind.com/geoip/geoip2/geolite2/ ")
+          @states[database_type].plugins.dup.each { |plugin| plugin.expire_action if plugin }
```

Comment:
(4) the actual expiry logic is in [plugin](https://github.com/logstash-plugins/logstash-filter-geoip/pull/181/files#diff-5421cf35b684cca803ff336c23f52918df0809819ae2bd6d36a4a4ec8250ddf1R169)